### DISCUSSION TITLE
Handling charts in different contexts

### DISCUSSION THREAD

**Discussion by jonbiemond** [sinaptik-ai/pandas-ai#277] [Ideas]
There are a lot of issues and discussions surrounding charts, I'd like to consolidate the discussion here. 

### The underlying issue
As I see it there are two closely related underlying issues, 1) the generated code does not always include a call to `plt.show()` making it hard for PandasAI to take plot-based behaviour, 2) users do not always want to have the charts displayed, specifically when using PandasAI as an API.

### Charts with PandasAI-as-an-API (#264, #166, #271)
When PandasAI is used as part of an app, it can be difficult for developers to capture the generated charts to pass to their apps. Currently it is possible to set `save_charts=True` and collect the chart from the output folder, however this isn't very elegant. The `streamlit` middleware handles this neatly (#257) by replacing the call to `plt.show()` with an expression that outputs the chart to `streamlit`. 

My concern is that it may not be very sustainable to create one-off solutions for every front-end request. Alternatively we could add a feature that returns the plot as an object instead of displaying it, this allows the developer to process the object as they see fit.

### Identifying plots in code (#205)
Currently PandasAI knows there are plots in the code by looking for calls to `plt.show()`. However the generated code does not consistently include `plt.show()` when creating plots. I think we will want to come up with a more reliable way to identify plots in the generated code.

Related:
- #264 
- #166 
- #221
- #257
- #271 
- #205

**Comment by jonbiemond** [comment]
Another recent issue (#281) where `plt.show()` is missing. Once we've figured out a way to consistently identify plots, we can implement a feature to "force" plots to display. I'm in favour of choosing a behaviour to be the default, and then users can toggle it depending on whether they want the charts to be displayed or returned as an object.

**Reply by gventuri** [reply]
That's a very good question. I don't have a strong idea of how to fix it unfortunately. One idea could be starting from the libraries that are imported in the process. It would already be a good starting point, but while it would fix the issue of matplotlib (#205), I cannot find a one-solution-fits all for all the charts libraries.

**Comment by eRuaro** [comment]
Hi @jonbiemond ! Is there currently a way to not have PandasAI block the code execution when using PandasAI through FastAPI when generating graphs?
